src/VisualStudio/Core/Def/LanguageService/AbstractCreateServicesOnTextViewConnection.cs

```diff
@@ -69,14 +73,29 @@ void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView text
         }
 
         private void InitializeServiceOnDocumentOpened(object sender, DocumentEventArgs e)
+            => InitializeServiceForDocument(e.Document);
+
+        private void InitializeServiceOnAlreadyOpenedDocuments(ImmutableArray<DocumentId> documentIds)
+        {
+            foreach (var id in documentIds)
+            {
+                var document = Workspace.CurrentSolution.GetDocument(id);
+                if (document != null)
+                {
+                    InitializeServiceForDocument(document);
+                }
+            }
+        }
+
+        private void InitializeServiceForDocument(Document document)
         {
-            if (e.Document.Project.Language != _languageName)
+            if (document.Project.Language != _languageName)
             {
                 return;
             }
 
             var token = _listener.BeginAsyncOperation(nameof(InitializeServiceForOpenedDocumentOnBackgroundAsync));
-            InitializeServiceForOpenedDocumentOnBackgroundAsync(e.Document).CompletesAsyncOperation(token);
+            InitializeServiceForOpenedDocumentOnBackgroundAsync(document).CompletesAsyncOperation(token);
```

Comment:
it feels like thsi is all bubbling down to `CompletionService.GetProjectCompletionProviders(document.Project)` which we need to do at most once per language.  so why not just collect the unique projects+languages up front and then load for all of those once on the bg? 

basically, i don't see why we're jumping to the bg for each document that may be open, instead of a single jump once.

Reply:
That's a good point.

Reply:
I have switched to a different approach. Now we don't try to warm up the cache when document is opened. Instead, we simply ignore providers from references if they are not already loaded and queue a task to load them in background. This would ensure the load never cause any delays, and in line with our general "best effort" approach for completion that prefers quick response over 100% correctness/completeness.